bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/I18nProviderImpl.java

```diff
@@ -0,0 +1,178 @@
+    private Locale locale;
+
+    // TranslationProvider
+    private ResourceBundleTracker resourceBundleTracker;
+
+    // LocationProvider
+    private static final String LOCATION = "location";
+    private PointType location;
+
+    @SuppressWarnings("unchecked")
+    protected void activate(ComponentContext componentContext) {
+        modified((Map<String, Object>) componentContext.getProperties());
+
+        this.resourceBundleTracker = new ResourceBundleTracker(componentContext.getBundleContext(), this);
+        this.resourceBundleTracker.open();
+    }
+
+    protected void deactivate(ComponentContext componentContext) {
+        this.resourceBundleTracker.close();
+    }
+
+    protected synchronized void modified(Map<String, Object> config) {
+        final String language = (String) config.get(LANGUAGE);
+        final String script = (String) config.get(SCRIPT);
+        final String region = (String) config.get(REGION);
+        final String variant = (String) config.get(VARIANT);
+        final String location = (String) config.get(LOCATION);
+
+        boolean locationChanged = false;
+        if (location != null) {
+            try {
+                PointType newLocation = PointType.valueOf(location);
+                if (newLocation != this.location) {
+                    locationChanged = true;
+                    this.location = newLocation;
+                }
+            } catch (IllegalArgumentException e) {
+                // preserve old location or null if none was set before
+                logger.warn("Could not set new location, keeping old one: ", e.getMessage());
+            }
+        }
+
+        if (StringUtils.isEmpty(language)) {
+            // at least the language must be defined otherwise the system default locale is used
+            logger.debug("No language set, fallback to default system locale");
+            locale = null;
+            return;
+        }
+
+        final Locale.Builder builder = new Locale.Builder();
+        try {
+            builder.setLanguage(language);
+        } catch (final RuntimeException ex) {
+            logger.warn("Language ({}) is invalid. Cannot create locale, keep old one.", language, ex);
+            return;
+        }
+
+        try {
+            builder.setScript(script);
+        } catch (final RuntimeException ex) {
+            logger.warn("Script ({}) is invalid. Skip it.", script, ex);
+            return;
+        }
+
+        try {
+            builder.setRegion(region);
+        } catch (final RuntimeException ex) {
+            logger.warn("Region ({}) is invalid. Skip it.", region, ex);
+            return;
+        }
+
+        try {
+            builder.setVariant(variant);
+        } catch (final RuntimeException ex) {
+            logger.warn("Variant ({}) is invalid. Skip it.", variant, ex);
+            return;
+        }
+
+        locale = builder.build();
+        if (locationChanged) {
```

Comment:
And what is if only location has changed, but not locale? This looks a bit unsymmetric to me here.